servers/services/src/main/java/org/projectnessie/services/hash/HashValidator.java

```diff
@@ -17,50 +17,94 @@
+  private BiConsumer<NamedRef, ParsedHash> validator = (namedRef, parsed) -> {};
 
-  /**
-   * Validates that a hash has been provided, and that it is unambiguous (that is, it starts with an
-   * absolute part).
-   */
-  HashValidator REQUIRED_UNAMBIGUOUS_HASH = REQUIRED_HASH.and(UNAMBIGUOUS_HASH);
+  public HashValidator() {
+    this("Hash");
+  }
+
+  public HashValidator(String hashDescription) {
+    this("Reference", hashDescription);
+  }
+
+  public HashValidator(String refDescription, String hashDescription) {
+    this.refDescription = refDescription;
+    this.hashDescription = hashDescription;
+  }
 
   /**
-   * Validates the provided hash.
+   * Validates the provided ref and hash.
    *
-   * @param name the name of the hash parameter, for error messages
-   * @param parsed the parsed hash, or {@code null} if no hash was provided
+   * @param namedRef the namedRef, required.
+   * @param parsed the parsed hash, or {@code null} if no hash was provided.
    */
-  void validate(String name, @Nullable @jakarta.annotation.Nullable ParsedHash parsed);
+  public void validate(
+      NamedRef namedRef, @Nullable @jakarta.annotation.Nullable ParsedHash parsed) {
+    validator.accept(namedRef, parsed);
+  }
+
+  /** Validates that a named ref is a branch. */
+  @CanIgnoreReturnValue
+  public HashValidator refMustBeBranch() {
+    validator =
+        validator.andThen(
+            (namedRef, parsed) ->
+                checkArgument(
+                    namedRef instanceof BranchName, "%s must be a branch.", refDescription));
+    return this;
+  }
+
+  /** Validates that a named ref is a branch or a tag. */
+  @CanIgnoreReturnValue
+  public HashValidator refMustBeBranchOrTag() {
+    validator =
+        validator.andThen(
+            (namedRef, parsed) ->
+                checkArgument(
+                    namedRef instanceof BranchName || namedRef instanceof TagName,
+                    "%s must be a branch or a tag.",
+                    refDescription));
+    return this;
+  }
+
+  /** Validates that a hash has been provided (absolute or relative). */
+  @CanIgnoreReturnValue
+  public HashValidator hashMustBePresent() {
+    validator =
+        validator.andThen(
+            (namedRef, parsed) ->
+                checkArgument(parsed != null, "%s must be provided.", hashDescription));
+    return this;
+  }
 
   /**
-   * Returns a new {@link HashValidator} that validates against both {@code this} and {@code other}.
+   * Validates that, if a hash was provided, it is unambiguous. A hash is unambiguous if it starts
+   * with an absolute part, because it will always resolve to the same hash, even if it also has
+   * relative parts.
    */
-  default HashValidator and(HashValidator other) {
-    return (name, parsed) -> {
-      validate(name, parsed);
-      other.validate(name, parsed);
-    };
+  @CanIgnoreReturnValue
+  public HashValidator hashMustNotBeAmbiguous() {
+    validator =
+        validator.andThen(
```

Comment:
It looks like top-level services use this method and `HashResolver` can also call this, which will cause duplicate conditions. Operation will still be correct, but it feels awkward... Would it be simpler to just keep several validation flags?